torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -393,15 +393,127 @@ def _assert_tensors_nonaliasing(inputs, outputs):
+    Get storage references from a tensor.
 
-    Returns all tracked tensor/symint variables that are not already in graph_output_vts.
+    Handles regular tensors, traceable wrapper subclasses, and sparse tensors.
+    Returns a set of StorageWeakRef objects.
+
+    Args:
+        tensor: The tensor to extract storages from
+
+    Returns:
+        Set of StorageWeakRef objects for the tensor's storage(s)
+    """
+    from torch._subclasses.fake_tensor import get_plain_tensors
+    from torch.multiprocessing.reductions import StorageWeakRef
+    from torch.utils._python_dispatch import is_traceable_wrapper_subclass
+
+    storages: set = set()
+
+    if not isinstance(tensor, torch.Tensor):
+        return storages
+
+    if tensor.is_sparse or tensor.is_sparse_csr:
+        return storages
+
+    if is_traceable_wrapper_subclass(tensor):
+        inner_tensors = []
+        get_plain_tensors(tensor, inner_tensors)
+        for inner in inner_tensors:
+            if isinstance(inner, torch.Tensor):
+                storages.add(StorageWeakRef(inner._typed_storage()))
+    else:
+        storages.add(StorageWeakRef(tensor._typed_storage()))
+
+    return storages
+
+
+class StorageAliasingTracker:
     """
+    Tracks storage references to detect aliasing between tensors.
+
+    This class encapsulates the logic for collecting storages from tensors
+    and checking for aliasing conflicts. Used to filter intermediate outputs
+    that would create input-output or output-output aliasing.
+    """
+
+    def __init__(self):
+        self.excluded_storages: set = set()
+
+    def _collect_storages_from_tensor(self, example_value):
+        """
+        Collect storage references from a tensor and add them to excluded storages.
+        """
+        self.excluded_storages.update(get_tensor_storages(example_value))
+
+    def collect_from_inputs(self, tx):
+        """Collect storages from graph input placeholders."""
+        from torch._higher_order_ops.utils import _collect_fake_inputs
+
+        for node in tx.output.graph.nodes:
+            if node.op == "placeholder":
+                example_value = _collect_fake_inputs([node])[0]
+                if isinstance(example_value, torch.Tensor):
+                    self._collect_storages_from_tensor(example_value)
+            else:
+                break
+
+    def collect_from_outputs(self, graph_output_vts):
+        """Collect storages from existing graph outputs."""
+        from torch._higher_order_ops.utils import _collect_fake_inputs
+
+        for vt in graph_output_vts:
+            proxy = vt.as_proxy()
+            example_value = _collect_fake_inputs([proxy.node])[0]
+            if isinstance(example_value, torch.Tensor):
+                self._collect_storages_from_tensor(example_value)
+
+    def check_and_track(self, proxy_node) -> bool:
+        """
+        Check if a proxy node can be added without aliasing, and track it if so.
```

Comment:
I dont know what "if a proxy node can be added without aliasing" means. Added to what without aliasing? Why are we tracking proxy nodes? I thought we were tracking storages.